File: python/algo/202603/algolcr132.py
class Solution:
    def cuttingBamboo(self, bamboo_len: int) -> int:
        mod = 10**9 + 7 
        ans = 1
        # 遍历分割段数k，从2到bamboo_len
        for k in range(2, bamboo_len + 1):
            frac = bamboo_len // k  # 整除结果（对应C++的/）
            rem = bamboo_len % k    # 余数
            res = 1
            # 计算rem个(frac+1)的乘积
            for _ in range(rem):
                res *= (frac + 1)
            # 计算(k-rem)个frac的乘积
            for _ in range(k - rem):
                res *= frac
            if res <= ans:
                break
            else:
                ans = res
        return ans % mod

    def cuttingBamboo2(self, bamboo_len: int) -> int:
        mod = 10**9 + 7 
        ans = 1
        
        for k in range(2, bamboo_len + 1):
            tot = bamboo_len
            res = 1
            while tot:
                div = (tot + k - 1) // k    # CONCRETE MATH gropus PAGE 97
                res *= div
                tot -= div
                k -= 1
            if res <= ans: break
            else: ans = res 
        return ans % mod   
    
    def cuttingBamboo3(self, bamboo_len: int) -> int:
        dp = [0] * (bamboo_len + 1)
        dp[0], dp[1] = 1, 1
        
        # 遍历每个长度i，从2到bamboo_len
        for i in range(2, bamboo_len + 1):
            # 遍历分割出的第一段长度len，从i-1递减到1
            len = i - 1
            while len > 0:
                current = max(dp[i - len], i - len) * len
                if current > dp[i]:
                    dp[i] = current
                len -= 1
        
        return dp[bamboo_len] % (10 ** 9 + 7)

File: cplus/leetcode/202603/algolcr132.cpp
#include <bits/stdc++.h>
using namespace std;

// LCR series
class Solution {
public:
    int cuttingBamboo(int bamboo_len) {
        // 动态规划 O(n^2)
        vector<int> dp(bamboo_len+1);
        dp[0] = 1; dp[1] = 1; // 至少分两段
        for (int i=2; i<=bamboo_len; i++) {
            for (int len=i-1; len; --len) {
                // max(dp[i-len],(i-len)) 剩下的 i-len 可以不拆分
                // 比如 dp[3] = 2, dp[6] = 3 * 3 > 3 * dp[3]
                dp[i] = max(dp[i], max(dp[i-len],(i-len)) * len);
            }
        }
        return dp[bamboo_len];

    }
    
    int cuttingBamboo2(int bamboo_len) {
        // 贪心
        const int mod = 1e9 + 7;
        auto pow = [] (long long base, int exp, int modulo) -> int {
            base %= modulo;
            long long res = 1;
            while (exp > 0)
            {
                if (exp & 1)
                    res = res * base % modulo;
                base = base * base % modulo;
                exp >>= 1;
            }
            return res;
        };

        if (bamboo_len <= 3) return bamboo_len - 1;
        int div = bamboo_len / 3, rem = bamboo_len % 3; 
        
        if (rem == 0) return pow(3, div, mod);
        else if (rem == 1) return 4LL * pow(3, div -1, mod) % mod;
        else return 2LL * pow(3, div, mod) % mod;        
    }

};